feat!: make initialize_or_replace take an option

nventuro opened the pull request:
Follow up on https://github.com/AztecProtocol/aztec-packages/pull/17017, this change allows for potentially cheaper circuits by not forcing the caller to compute the initial value (which might be expensive) unless needed. The multiple `Option` methods such as `unwrap_or`, `unwrap_or_else`, etc., make handling this simple.

E.g. a counter becomes
```noir
storage.counter.initialize_or_replace(|opt_current| UintNote::new(
  opt_current
    .map(|note| note.value)
    .unwrap_or(0) + 1
));
```

benesjan commented on noir-projects/aztec-nr/aztec/src/state_vars/private_mutable.nr:
```diff
@@ -263,12 +263,13 @@ where
     /// This function can only be called after the PrivateMutable has been initialized.
     /// If called on an uninitialized PrivateMutable, it will fail because there is
-    /// no current note to replace.
+    /// no current note to replace. If you don't know of the state variable has been
```
```suggestion
    /// no current note to replace. If you don't know if the state variable has been
```

benesjan reviewed (approved):
Nice

nventuro committed:
feat!: make initialize_or_replace take an option

Follow up on https://github.com/AztecProtocol/aztec-packages/pull/17017, this change allows for potentially cheaper circuits by not forcing the caller to compute the initial value (which might be expensive) unless needed. The multiple `Option` methods such as `unwrap_or`, `unwrap_or_else`, etc., make handling this simple.

E.g. a counter becomes
```noir
storage.counter.initialize_or_replace(|opt_current| UintNote::new(
  opt_current
    .map(|note| note.value)
    .unwrap_or(0) + 1
));
```
